jobs/video_jobs/maya_submit.sh.erb

```diff
@@ -12,62 +12,118 @@ BIND_DIRS="$JOB_TMPDIR:/tmp"
-
-    convert $IMG_DIR/$IMG $IMG_ARGS $THUMB_DIR/$IMG_NAME.png
-  done
-}
-
-function start_frame(){
-  if [ -z "$PBS_ARRAYID" ]; then
-    PBS_ARRAYID=1
-  fi
-  
-  if [ $PBS_ARRAYID -gt 1 ]; then
-    # if nodes >1 have to offset +1 so the start frame of this task
-    # doesn't overlap with the end frame of the previous task
-    TASK_START_FRAME=$(( OFFSET - TASK_SIZE + START_FRAME + 1 ))
-  elif [ "$NODES" != "1" ]; then
-    TASK_START_FRAME=$(( OFFSET - TASK_SIZE + START_FRAME ))
-  else
-    TASK_START_FRAME=$START_FRAME
-  fi
+UTIL_FILE="$JOB_TMPDIR/utils.rb"
 
-  echo $TASK_START_FRAME
-}
+cat > "$UTIL_FILE" <<UTILS
+#!/usr/bin/env ruby
 
-function end_frame(){
+def output_start_frame
+  puts task_start_frame
+end
 
-  if [ $REMAINDER -ne 0 ] && [ "$PBS_ARRAYID" -eq "$NODES" ]; then
-    # last task picks up the remainder of the frames
-    TASK_END_FRAME=$(( OFFSET + START_FRAME + REMAINDER ))
-  elif [ "$NODES" != "1" ]; then
-    TASK_END_FRAME=$(( OFFSET + START_FRAME ))
+def task_start_frame
+  if nodes == 1
+    start_frame
+  elsif task_size == 1
+    # you have as many nodes as there are tasks
+    array_id
+  elsif array_id > 1
+    start_frame + offset
   else
-    TASK_END_FRAME=$END_FRAME
-  fi
-
-  echo $TASK_END_FRAME
-}
+    start_frame
+  end
+end
+
+def output_end_frame
+  puts task_end_frame
+end
+
+def task_end_frame
+  if nodes == 1
+    end_frame
+  elsif task_size == 1
+    # you have as many nodes as there are tasks
+    array_id
+  elsif array_id > 1 && array_id == nodes
+    # last task gets the remaining frames
+    offset + task_size + remaining_frames
+  elsif array_id > 1
+    offset + task_size
+  else
+    # array id == 1
+    task_size
+  end
+end
+
+def array_id
+  (ENV['PBS_ARRAYID'] || 1).to_i
+end
+
+def offset
+  (array_id - 1) * task_size
+end
+
+def task_size
+  task_size = (total_frames / nodes).to_i
+  task_size.zero? ? 1 : task_size
+end
+
+def total_frames
+  # need +1 here bc we're always starting at the first frame
+  # 20 - 1 = is 20 frames, not 19. even 3-3 is 1 frame (the third)
+  end_frame - start_frame + 1
+end
+
+def nodes
+  (ENV['NODES'] || 1).to_i
+end
+
+def start_frame
+  frame = (ENV['START_FRAME'] || 1).to_i
+  frame <= 0 ? 1 : frame
+end
+
+def end_frame
+  (ENV['END_FRAME'] || 1).to_i
+end
+
+def remaining_frames
+  total_frames % nodes
+end
+
+def create_thumbnails
+  img_dir="<%= project_dir %>/images"
+  thumb_dir="<%= project_dir %>/thumbnails"
+
+  images = Dir.glob("#{img_dir}/*").keep_if do |img|
+    file_num = /\d+/.match(File.basename(img))
+    if !file_num.nil?
+      num = file_num.to_s.to_i
+      num >= task_start_frame && num <= task_end_frame
+    else
+      false
+    end
+  end
+
+  convert_args="-resize 80x80"
+  Dir.mkdir(thumb_dir) unless File.exists?(thumb_dir)
+
+  images.each do |img|
+    fname = File.basename(img, ".*")
+    puts "creating thumbnail for #{fname}"
+    system("convert #{img} #{convert_args} #{thumb_dir}/#{fname}.png")
+  end
```

Comment:
I'll look into escaping this since the image names may have spaces or other strange characters.

Reply:
I'll have this in another PR.